src/lib/storages/RedisProxyCache.ts

```diff
@@ -0,0 +1,171 @@
+import Redis from 'ioredis';
+// think, how to avoid direct deps on Redis class (move to a separate fn?)
+
+import * as validation from '../../validation';
+import { IProxyCache, RedisProxyCacheConfig, IsLastFailure, AlsRequestDetails } from '../../types';
+import { BasicProxyCache } from './BasicProxyCache';
+import { REDIS_KEYS_PREFIXES, REDIS_SUCCESS, REDIS_IS_CONNECTED_STATUSES } from './constants';
+
+export class RedisProxyCache extends BasicProxyCache<RedisProxyCacheConfig> implements IProxyCache {
+  private readonly redisClient: Redis;
+
+  constructor(proxyConfig: RedisProxyCacheConfig) {
+    super(proxyConfig);
+    this.redisClient = this.createRedisClient(proxyConfig);
+  }
+
+  async addDfspIdToProxyMapping(dfspId: string, proxyId: string): Promise<boolean> {
+    const key = RedisProxyCache.formatDfspCacheKey(dfspId);
+    const response = await this.redisClient.set(key, proxyId);
+    const isAdded = response === REDIS_SUCCESS;
+    this.log.debug('proxyMapping is added', { key, proxyId, isAdded });
+    return isAdded;
+  }
+
+  async lookupProxyByDfspId(dfspId: string): Promise<string | null> {
+    const key = RedisProxyCache.formatDfspCacheKey(dfspId);
+    const proxyId = await this.redisClient.get(key);
+    this.log.debug('lookupProxyByDfspId is done', { key, proxyId });
+    return proxyId;
+  }
+
+  async removeDfspIdFromProxyMapping(dfspId: string): Promise<boolean> {
+    const key = RedisProxyCache.formatDfspCacheKey(dfspId);
+    const result = await this.redisClient.del(key);
+    const isRemoved = result === 1;
+    this.log.debug('proxyMapping is removed', { key, isRemoved, result });
+    return isRemoved;
+  }
+
+  async setSendToProxiesList(alsReq: AlsRequestDetails, proxyIds: string[], ttlSec: number): Promise<boolean> {
+    const key = RedisProxyCache.formatAlsCacheKey(alsReq);
+
+    const isExists = await this.redisClient.exists(key);
```

Comment:
Unfortunately, there's no `setIfExists` functionality in Redis with sets.
Also if we have several parallel requests with the same sourceId/type/partyId and _proxyIdsList_, it won't be a problem, coz  only one of them will be resolved with _true_, meaning we can send discovery calls for all proxies.
The issue we might have with current impl. is when we have parallel requests with different _proxyIdsList_. To address it we need to use Lua scripts, I think.
Will check how it could be implemented